### ISSUE TITLE
Use a plugin system to refactor the life cycle of a PyScript page

### ISSUE THREAD

**Issue by antocuni** [pyscript/pyscript#763]
TL;DR: I'm proposing to model the life cycle of a PyScript page using a plugin system. The main logic is very slim, it calls hooks at well defined times, and most of the functionality is implemented inside plugins, some of which are builtin. This is loosely modeled after what e.g. `pytest` does, but I'm sure there are other systems following a similar model

In order to design the new system, we need to first analyze what is the current flow of execution. To avoid having a giant wall of text, I'm going to analyze the current status in this post, and write a proposal for the refactoring in the next one.

**Current control flow**

1. The HTML page is parsed. We encounter `<script defer src=".../pyscript.js>` Because of `defer`, the script is executed when the document is fully parsed.

2. pyscript.js is executed. We start from `main.ts`: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/main.ts#L1-L43

3. main.ts imports a bunch of modules; in particular, `pyscript.ts` has also some module-level statements which call `addInitializer` and `addPostInitializer`:
https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pyscript.ts#L265-L266

4. we call `customElements.define` for all CE. As soon as each of them is defined, *all* the tags in the document are connected. This means that the order of execution of the callbacks might be different than the order of the tags in the HTML. Related issue: #761.
*Note*: that at this point in time nothing else is loaded yet; no runtime, no `py-config`, no `py-env`, etc.
Scripts are enqueued but not executed.
https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/main.ts#L19-L28

5. The various connectedCallback are called. In the following, we say "DOM logic" to mean all the logic which aims at locating/adding/modifying elements in the DOM, e.g. to create the `<div>` to hold the output of a script:

     - PyScript: some immediate DOM logic, then it calls `addToScriptsQueue`
     - PyRepl: only immadiate DOM logic
     - PyEnv: very few logic done immediately; it uses `addInitializer` to call `loadEnv` and `loadPaths` later
     - PyBox, PyTitle: only immadiate DOM logic
     - PyButton, PyInputBox: immediate DOM logic + `this.runAfterRuntimeInitialized` (which uses `runtimeLoaded.subscribe`)
     - PyWidget: after a bit of indirections, it ends up calling
       `runtimeLoaded.subscribe`, but then in the callback we use a 1sec delay
       before actually instantiating `proxyClass` (which is a python class)
     - PyConfig: see later
  
6. after the imports, there is some core logic in main.ts: we create a `<py-config>` element if needed. So precise timing at which `PyConfig.connectedCallback` happens depends on whether there is a real tag or not:

     - if there was a `<py-config>` tag, it was called at point (4)
     - else, it is called now.

7. `pyconfig.connectedCallback` calls loadRuntimes(), which create a `<script>` tag to download pyodide, and add an event listener for its `load()` event:
https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pyconfig.ts#L62-L73

8. we continue the main flow of execution: in main.ts we add the `<py-loader>` element to the DOM; the splashscreen is showed only now. main.ts finally finishes execution:
https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/main.ts#L39-L43

9. eventually the `<script>` tag added at point (7) finishes loading, and it fires the `load` event (which is implemented inside `pyconfig.ts:loadRuntimes`). We call `runtimeObj.initialize()`.  The execution flow bounces back and forth between the base class in runtime.ts and the concrete subclass in `pyodide.ts`. We call `loadInterpreter`.
https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/runtime.ts#L135-L181

10. We call all the things which were enqueued earlier. We start with `initializers`, which currently are:

    - `pyscript.ts:mountElements`: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pyscript.ts#L254-L265

    - `pyenv.ts:loadEnv` and `loadPaths`: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pyenv.ts#L59-L79

11. we execute `<py-script>` tags: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/runtime.ts#L161-L165

12. We close the loader (why so early?): https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/runtime.ts#L170-L172

13. we run `postInitializers`. Currently the only one is `initHandlers`: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pyscript.ts#L213-L219

14. at some point, we also run the callbacks registered by using `runAfterRuntimeInitialized`, which currently is invoked by:
    - pybutton.ts
    - pyinputbox.ts
    - in both cases above, what it does is to run further user-define code inside e.g. the `<py-button>` tag, but using a different way than for `<py-script>`: https://github.com/pyscript/pyscript/blob/6023c413ab1762e0a1bb2476ed46eb4be2af7944/pyscriptjs/src/components/pybutton.ts#L69-L73

**Comment by antocuni** [comment]
**Proposal for a refactoring**

The execution logic described above is a bit messy and it's not obvious to follow what's going on. Let's try to determine a high-level summary of what we want to do:
   1. load the config
   2. determine the runtime, add a `<script>` to featch/load/initialize it. We want to do this is early as possible to minimize the latency
   3. open the splashscreen
   4. define all custom elements
   5. wait until the runtime is ready
   `--- runtime ready ---`
   6. install the env
   7. import all the python plugins, and call `pyscript_init` on each of them
   8. execute all the user-defined scripts in the right order
   9. run what we currently call `post initializers`
  10. close the splashscreen

As said above, the idea is to have a very slim `main`, and delegate as much as possible to plugins.
The following is mostly pseudocode but gives the idea of what I'd like to do. 
To implement a plugin you subclass `Plugin` and override the desired hooks:
```typescript
class Plugin extends Object {

    configure(config) {
        /* give the plugins early access to the config. Potentially, they can
           modify it, and modifications will be visible to later steps.

           Example of things which plugins might want to do:
             - early sanity check about their own options
             - rename/remap some options
             - add new packages to install
             - register new runtime engines
             - modify options for other plugins (e.g.: a 'debugger' plugin
               might set the option 'always_show_the_terminal' or something
               like that)
        */
    }

    start(config) {
        /* main entry point for plugins. At this point, config should be
           considered finalized.

           Examples:
             - define custom elements
             - start fetching external resources
        */
    }

    runtime_setup(config, runtime) {
        /* called once the runtime is ready to execute commands

           Examples:
             - pip install packages
             - import/initialize python plugins
        */
    }

    before_evalute(config, runtime) {
        /* Everything is ready.

           Called before evaluating <py-script> tags.
        */
    }

    after_evaluate(config, runtime) {
        /* <py-script> tags have been evaluated

           Examples:
             - close the splashscreen
        */
    }
}
```

With this, the main logic boils down to:
1. load the config
2. load the plugins
3. calls the hooks at the right time

```typescript
main() {
    // CURRENT ASSUMPTION: the page has already been loaded because we use
    // <script defer ...>

    // builtin plugins, providing the core of functionalities
    pyscript_register_plugin(BuiltinLogic);
    pyscript_register_plugin(PyodidePlugin);
    pyscript_register_plugin(Splashscreen);

    // 1. load the config
    const config = load_config(); // look up and parse `<py-config>`
    const plugins: Plugin[] = load_plugins(config);

    for(const p of plugins)
        p.configure(config);

    // 2. determine the runtime, add a <script> to featch/load/initialize it
    let runtime_name = config.runtime;
    let runtime_promise = create_runtime(runtime_name);

    for(const p of plugins)
        // 3. open the splashscreen
        // 4. define custom elements
        p.start(config);

    // 5. wait until the runtime is ready
    let runtime = await runtime_promise;

    for(const p of plugins)
        // 6. install the env
        // 7. import/init python plugins
        p.runtime_setup(config, runtime);

    let scripts = magically_get_the_list_of_pyscript_tags();
    for(const s of scripts)
        // 8. execute all the scripts in the right order
        s.evaluate();

    for(const p of plugins)
        //  9. the equivalent of current postInitializers
        // 10. close the splashscreen
        p.after_evaluate();
}
```

**Comment by antocuni** [comment]
**Examples of plugins**
The following code shows more or less how to implement the current logic with the plugin system defined above:
```typescript
class PyodidePlugin extends Plugin {
    configure(config) {
        // PyodideRuntime is currently defined in pyodide.ts
        pyscript_register_runtime("pyodide", PyodideRuntime);
    }
}

class BuiltinLogic extends Plugin {

    start(config) {
        customElements.define('py-script', PyScript);
        customElements.define('py-box', PyBox);
        customElements.define('py-button', PyButton);
        customElements.define('py-title', PyTitle);
        ...
    }

    runtime_setup(config, runtime) {
        runtime.installPackage(config.packages);

        // import and initialize python plugins
        for(const name in get_py_plugins(config)) {
            runtime.run(`
                import ${name}
                ${name}.pyscript_init()
            `)
        }
    }

    after_evaluate(config, runtime) {
        // currently defined in pyscript.ts
        initHandlers();
    }
}

class Splashscreen extends Plugin {

    start(config) {
        logger.info('add py-loader');
        this.loader = <PyLoader>document.createElement('py-loader');
        document.body.append(loader);
    }

    after_evaluate(config, runtime) {
        this.loader.close();
    }
}
```

**Comment by antocuni** [comment]
**Open questions**

- sync or async? The code above is mostly sync, apart when we wait for the runtime to fully load. Maybe it's a good idea to declare all hooks as `async`, start of of them asynchronously and then `await` all before moving to the next hook?

- defer or not defer? The code above assumes that `main()` is executed after the page has been loaded, thanks to `defer`. We might want to revisit this decision and use a different approach, so that pyscript can start downloading and initializing pyodide very soon (especially if we put it in a webworker), and then wait until the page loads to do the more expensive sync operations like executing `<py-script>` tags.

- order of execution of different plugins. Currently this is left unspecified, but we might want to have a system to declare priority of execution. E.g., for the `Splashscreen` we would that its `start()` is called before all the others (to show the splash screen as early as possible) and its `after_evaluate()` is called as late as possible.

**Comment by FabioRosado** [comment]
This is both a great proposal and a great explanation of the execution flow! I've definitely been thrown off a few times when working on the PR to move the custom elements to its own module.

Quick question/comment:

```
    for(const p of plugins)
        // 3. open the splashscreen
        // 4. define custom elements
        p.start(config);

    // 5. wait until the runtime is ready
```

Should we wait for runtime before we start iterating over the plugins? I'm asking because this will likely throw that "run in undefined" since `runAfterRuntimeInitialized` will be called when we create the custom element.

I guess this could be solved by specifying the execution order of different plugins (+1 to that).

Regarding the sync/async, I think keeping them async may be a good idea as well

**Comment by ntoll** [comment]
Epic write up @antocuni - a really useful summary.

Questions/points:

* Why is the splash screen not the first thing we do (i.e. giving user some indication that something is happening)..?
* It's not obvious that `configure` comes before `start` (which feels like it should come first) in the `Plugin` class. How about `start` -> `initialise` ..? The others seem to make sense.
* I like very much the clear ordering of steps... `runtime_setup` is guaranteed to come before `before_evaluate` which is guaranteed to ensure `after_evaluate` only happens when the scripts have been evaluated.
* Regarding your open questions... my intuition is:
  * Since we're in the world of the browser when undergoing this process, be as browser-y as possible, so be async and await.
  * Currently defer means things are simple in terms of order of activity. I suggest we just make it work as is, and then explore a more complicated (non-defer) way to do things.
  * Yes to your suggestion to declare priority of execution of plugins. I can see lots of ways in which this would be helpful to end users.
* What about web-workers? I mean, the examples you give assume we're running in the browser's main thread. How do you see the web-worker story fitting into what you've outlined.

I really like your approach. I wonder if there are aspects of the browser's "mechanics" (message bus, custom events etc...) we could incorporate into this. I suspect @tedpatrick will have a deeper understanding of these things and be able to shed light on how they might be brought together to our advantage.

What I especially like is:

* This is simple to understand (and thus maintain). I'd love a small, simple **core**.
* There are clear separations of concern into plugins. A pattern that's well understood.
* We're able to guarantee order or execution. Making it easier for folks to think about as they build applications upon PyScript. Don't make me think (too much)..! ;-)

Bravissimo..!

**Comment by tedpatrick** [comment]
The browser has an **excellent** event system. It is a very effective way to handle plugin-like systems and the api is very small, supports custom events, and you have likely already used it via addEventListener.

Event Phases
- Capture - Event from root to target. Think: events arriving
- Target - Event at the target. Think: events delivered at target
- Bubble - Event from target to root. Think: events leaving

You can listen for events arriving/target/leaving on any element of the DOM.

This is an ideal way for these "plugin" parts to coordinate and message one another in using these 3 methods:
addEventListener() # setup a listener, this is a subscription on an element for an event denoting phase
dispatchEvent() # emit an event into the dom
removeEventListener() # remove a listener, this unsubscribes on an element.

py-config
 - root.dispatchEvent( new Event("pyconfig:loading",{"bubbles":true, "cancelable":false} ) )
 - root.dispatchEvent( new Event("pyconfig:loaded", {"bubbles":true, "cancelable":false} ) )

So we have a way for these elements to communicate, share data/events and wire things together.

**Comment by antocuni** [comment]
Let me answer to various questions one by one.

reply to @FabioRosado 

> Should we wait for runtime before we start iterating over the plugins? I'm asking because this will likely throw that "run in undefined" since `runAfterRuntimeInitialized` will be called when we create the custom element.

My idea is that some plugins might want to run before the runtime is initialized (e.g., the splashscreen). But note that not all the hooks receives the `runtime` argument: when you receive it, you can be sure it is ready

---

reply to @ntoll 

> * Why is the splash screen not the first thing we do (i.e. giving user some indication that something is happening)..?

My idea is that we want to start downloading pyodide as soon as possible, and thus we do it before calling the `start` hook (which will the main entry point of all plugins, so potentially costly). Note that downloading pyodide is async and non-blocking, so the splash screen is displayed very quickly anyway.
I thought about introducing *yet another* hook like `pre_start`, which would be called earlier, but it didn't seem worth it.


> * It's not obvious that `configure` comes before `start` (which feels like it should come first) in the `Plugin` class. How about `start` -> `initialise` ..? The others seem to make sense.

this is modeled after pytest, whose [`pytest_configure`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_configure) hook is called very early, and before e.g. `pytest_startsession`.
The idea is that if you want to modify the config, you want to do it very early before the other plugins start.

> * Regarding your open questions... my intuition is:
>   
>   * Since we're in the world of the browser when undergoing this process, be as browser-y as possible, so be async and await.

yes, that would be my gut feeling as well. But on the other hand, async/await is useful *only* if you expect blocking calls (such as networking). If you execute only CPU-bound code, async/await introduce complexity and non-linear flow which might not be worth it.
So maybe the question is: in which of the hooks we can anticipate having network-bound code?

>   * Currently defer means things are simple in terms of order of activity. I suggest we just make it work as is, and then explore a more complicated (non-defer) way to do things.

+1

> * What about web-workers? I mean, the examples you give assume we're running in the browser's main thread. How do you see the web-worker story fitting into what you've outlined.

that's a great question. I didn't think too much about it: one possibility is to hide the complexity of the web worker under the `Runtime` class: i.e., you call `runtime.run` as now but it sends a message to the web worker under the hood. But maybe it's better to make it more explicit.

---

Reply to @tedpatrick 

> The browser has an **excellent** event system. It is a very effective way to handle plugin-like systems and the api is very small, supports custom events, and you have likely already used it via addEventListener.
[cut]

wow, thins looks very interesting, I like the idea! Thanks!
My immediate question is: how do we sync between the various phases?
In your example, how can I want until all the handlers for `pyconfig:loading` has been called and returned?

**Comment by tedpatrick** [comment]
Reply to @antocuni 

> how do we sync between the various phases?
The different phases allow you to listen in different locations. Sometimes you want to capture as events go down the DOM towards the target (there is an API for canceling events, if cancellable), sometimes you are only interested in events at the target (click), and sometimes you want to capture the event as it returns up the DOM tree.

> In your example, how can I wait until all the handlers for pyconfig:loading have been called and returned?
I assume py-config can calculate the things to load and eventually capture all child dependencies of dependencies. As the items load the loader can emit progress events and track the loading set to completion. The loading screen might render progress indeterminate until the known set of loaders is known and show what is loading in the text on the screen. Events can contain data payloads and many things can subscribe to this stream of events.

```javascript
root.dispatchEvent( new CustomEvent('found', { detail: { type: 'cat', owner:'fabio' } } ) )
```

So rather than hooks, you could just have plugins subscribe, emit events, and use events to coordinate behavior. This system is also very well worn in it is the primary way elements interact. 

**mouse events** ([MouseEvent](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent)): mousedown, mouseup, click, dblclick, mousemove, mouseover, mousewheel, mouseout, contextmenu
**touch events** ([TouchEvent](https://developer.mozilla.org/en-US/docs/Web/API/TouchEvent)): touchstart, touchmove, touchend, touchcancel
**keyboard events** ([KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent)): keydown, keypress, keyup
**form events**: focus, blur, change, submit
**window events**: scroll, resize, hashchange, load, unload

**Comment by ntoll** [comment]
My 2c.

Can we marry the simplicity of @antocuni's "flow" and the inherent idioms of the browser (a la @tedpatrick's description of events)..? I suspect this would involve defining new `CustomEvent`s to which other things subscribe. My biggest concern with this approach, though, is the asynchronous nature of how things unfold... it feels like we need a deferred/future/promise to resolve when all the `n` (where we cannot know the value of `n`) number of handlers for `foo` have finished `bar`-ing thus allowing us to signal a `baz` event.

**Comment by fpliger** [comment]
Thank you @antocuni  this is an **excellent** write up and so useful in so many levels! I'll split my comments separately to try to avoid mixing concerns (I already know I'll fail but, hey, one can try!).

The way I see there are 4 layers to this:

1. Current status/analysis of the flow and finding areas of opportunities to improve (I'll provide feedback to each item that needs it below)
2. How do we improve these things proposal
3. Agree on a higher level API for plugins
4. Internally design and make sure that point 2 supports point 3

Obviously there 5, which is to go implement plugins

So...

## Current status/analysis of the flow and finding areas of opportunities to improve 

> 1. The HTML page is parsed. We encounter <script defer src=".../pyscript.js> Because of defer, the script is executed when the document is fully parsed.

Yeah, this sucks :) 

It's been a known tech debt since day 1. We need to remove the `Svelte` dependency here. I've removed 95% of it and only think missing is the `stores`. @tedpatrick  has a POC of a potential replacement so I'm hopeful!

> main.ts imports a bunch of modules; in particular, pyscript.ts has also some module-level statements which call addInitializer and addPostInitializer:

Yes, `pyscript` happens to be the only one now but it doesn't necessarily mean it'll be the only one forever

> we call customElements.define for all CE. As soon as each of them is defined, all the tags in the document are connected.

Yep. The order is wrong, for sure

> after the imports, there is some core logic in main.ts: we create a <py-config> element if needed. So precise timing at which PyConfig.connectedCallback happens depends on whether there is a real tag or not:
> 
> if there was a <py-config> tag, it was called at point (4)
> else, it is called now.

Yep. Config should go first.

> We close the loader (why so early?)

Initially, when talked about post_initializers these were for things to run after the main app log was rendered (by running the scripts) so releasing the loader earlier gives the user the impression that the app is read sooner (which, we all know, now this sooner is way tooooo slow).

MOre about this later but, imho, the "right" point is not when we decide to close the loader but to actually give users full control IF to show the loader at all, when to close it and to control it's aspect and how invasive it is. The "loader" should really just be a set of hooks that provide users ways to control the interface and then we can give some Plugins that do things in different ways. For instance, have one loader that is basically a splashscreen that takes over the whole screen like now, another one that show a spinner placeholder on every <py-script> or element waiting, or just a small loading icon on the bottom/top corner of the page, or..... whatever. Point is that there are N ways users may want it and we should allow freedom.


In addition to the points above, I think the PyWidget execution (that actually registers custom extensions) is missing. It's actually accessing the [runtime directly to do that](https://github.com/pyscript/pyscript/blob/main/pyscriptjs/src/components/base.ts#L245-L257), after it has loaded.

Again, great analysis! Thank you!

... more below

**Comment by fpliger** [comment]
## Proposal for refactoring

> open the splashscreen

As I hinted above, I'd like this to not be the default behaviour but to:

1. have a less invasive default behavior
2. allow plugins/users to fully customize and control this

> import all the python plugins, and call pyscript_init on each of them

+1 on ~~import~~ load python plugin (`pyscript_init` is an implementation detail and depends on the hooks we will define)

> execute all the user-defined scripts in the right order

do you mean only `<py-script>` tags? or you include anything that actually defines code that needs to be executed?

> close the splashscreen

to be consistent with my comment above, IMO this should be a hook

**Comment by fpliger** [comment]
## Plugins API proposal

I'll be honest and start with saying that I'm confused. I thought we've been talking about Python plugins but yet the proposal, in the issue is not. I mean, we always said we should accept both types, Python and JS plugins so not that the one negate the other but I thought we specified we wanted to start with Python support (even if [a little, for some definition of little 😉 ] harder )

Anyway, for TS/JS plugins I'm +1 of subclassing as a method of defining a Plugin. 

I don't entirely get all the hooks there and the use case, but I think we can go over them with real examples and how they could be useful for real use today. In general, I'd prefer to start with a small subset that we know are needed/matter and then expand what's supported later. YAGNI ftw. 


Now, if we are talking about **Python plugins** I strongly think that we should support definition plugins in a more pythonic way. I can think of 2 options:

1. class based plugins 
2. function based plugins

Both have implementation constraints related to how to pass Py<->Js objects around and I will skip implementation details for now.

I'd like for us to support both 1 and 2 above actually and allow users to define easy and quick plugins using 2 and more complex with 1.

Maybe we should have 2 separate issues: this one to define JS plugins and a new one to discuss Python plugins  API  (with the hooks through the App lifecycle being shared across the 2

**Comment by pauleveritt** [comment]
My two cents: I agree with @fpliger ... I think the lifecycle on the JS should be the minimum.

I confess to a secret agenda: I'm hoping to prove we can write custom elements completely in Python. I recognize that's a colossal change. I doubt PyScript will look at it and I'll do it as a Pyodide app. But in that mode, you want the "source of truth" in the interpreter.

**Comment by pauleveritt** [comment]
Now, a more constructive response to @antocuni and his work. 😀 I have "Diary Driven Development" showing your 1-5 ("load config" to "wait until the runtime is ready") in a dirt-simple fashion:

- Have a well-known data attribute added on startup to the body `<body data-pslifescyle>`
- As things change, change the value
- Have Mutation Observers that watch this and act

I have a feeling that, if we limit bootstrapping to something simple, we could get all the goals, just web platform stuff, no Svelte Store.

**Comment by antocuni** [comment]
wow, lots of things to answer. Let's start from @fpliger comments:

> ## Current status/analysis of the flow and finding areas of opportunities to improve
> > 1. The HTML page is parsed. We encounter <script defer src=".../pyscript.js> Because of defer, the script is executed when the document is fully parsed.
> 
> Yeah, this sucks :)

good! I'll try get rid of it!

> It's been a known tech debt since day 1. We need to remove the `Svelte` dependency here. I've removed 95% of it and only think missing is the `stores`. @tedpatrick has a POC of a potential replacement so I'm hopeful!

maybe it was not clear from my words, but my proposal should also remove the dependency on svelte stores: if you look, there is no global state and all plugins receive what they need (the config and the runtime) as parameters.

I *think* this is enough to kill svelte stores, unless I am missing something (which is entirely possible).

> > main.ts imports a bunch of modules; in particular, pyscript.ts has also some module-level statements which call addInitializer and addPostInitializer:
> 
> Yes, `pyscript` happens to be the only one now but it doesn't necessarily mean it'll be the only one forever

+1. My proposal supports this use case.

> > We close the loader (why so early?)
> 
> MOre about this later but, imho, the "right" point is not when we decide to close the loader but to actually give users full control IF to show the loader at all, when to close it and to control it's aspect and how invasive it is. 

+1. In my analysis above I just written down what is the current behavior of the loader and added enough hooks to be able to replicate it. But once it is a plugin it will be very simple to disable it and/or swap it with another plugin and/or to configure its behavior from `<py-config>`.

> In addition to the points above, I think the PyWidget execution (that actually registers custom extensions) is missing. It's actually accessing the [runtime directly to do that](https://github.com/pyscript/pyscript/blob/main/pyscriptjs/src/components/base.ts#L245-L257), after it has loaded.

right, I didn't analyze it separately. But I think it already fits nicely the plugin workflow which I proposed above.


> ## Plugins API proposal
> I'll be honest and start with saying that I'm confused. I thought we've been talking about Python plugins but yet the proposal, in the issue is not. I mean, we always said we should accept both types, Python and JS plugins so not that the one negate the other but I thought we specified we wanted to start with Python support (even if [a little, for some definition of little wink ] harder )

right, I might have been unclear with my intentions.
My reasoning was the following:
1. the current workflow is a mess and needs to be refactored
2. I designed in my head a possible internal refactoring
3. while thinking of (2), I thought it could be the basis for the JS-based plugin system
4. finally, once we have a clear execution flow, we can easily add a hook to execute python plugins.

The alternative is to start with (4), but then it would be to create yet another ad-hoc hook which makes (1) even more complicated. So, overall it's probably less work to do (1) and (2) first.

After the feedback which I got, I am less sure that it's a good idea to do (3) right now. Probably it's a better idea to do the internal refactoring but do not publicize it as a way to develop external plugins: this will give us the possibility of playing with it internally, understand whether it's a good system and then we can decide whether we want to expose it also to the external world.

> Anyway, for TS/JS plugins I'm +1 of subclassing as a method of defining a Plugin.

+1

> I don't entirely get all the hooks there and the use case, but I think we can go over them with real examples and how they could be useful for real use today. In general, I'd prefer to start with a small subset that we know are needed/matter and then expand what's supported later. YAGNI ftw.

yes, see my comment above. I think we need to start with some set of hooks, but probably is a good idea to wait a bit before writing it into stone.

> Now, if we are talking about **Python plugins** I strongly think that we should support definition plugins in a more pythonic way. I can think of 2 options:
[cut]

I'm not talking about Python plugins here :)

> Maybe we should have 2 separate issues: this one to define JS plugins and a new one to discuss Python plugins API (with the hooks through the App lifecycle being shared across the 2

Exactly!
As the title says, the main goal of this issue is to refactor the life cycle of the page.
I think that Python plugins would fit this proposed model very well, but better to talk about them separately.

**Comment by antocuni** [comment]
> My two cents: I agree with @fpliger ... I think the lifecycle on the JS should be the minimum.

I don't understand: what do you mean with that?

> I confess to a secret agenda: I'm hoping to prove we can write custom elements completely in Python. I recognize that's a colossal change. I doubt PyScript will look at it and I'll do it as a Pyodide app. But in that mode, you want the "source of truth" in the interpreter.

personally I'm happy to move more logic to Python but we need to be aware of the tradeoffs:
1. at this moment in time, python is slower (might be good enough for some cases, not for others)
2. there are things which you cannot do in Python because you want to do them *before* the runtime is initialized

> Now, a more constructive response to @antocuni and his work. grinning I have "Diary Driven Development" showing your 1-5 ("load config" to "wait until the runtime is ready") in a dirt-simple fashion:
> 
> * Have a well-known data attribute added on startup to the body `<body data-pslifescyle>`
> * As things change, change the value
> * Have Mutation Observers that watch this and act

I like this idea, but it solves only a fraction of the issue. This is basically a pub/sub mechanism which is good e.g. for the splashscreen but it doesn't allow an easy way to sync between the various activities (e.g.: I want to wait until all the plugins have initialized the `runtime`)

> I have a feeling that, if we limit bootstrapping to something simple, we could get all the goals, just web platform stuff, no Svelte Store.

+1 to kill svelte stores :)

**Comment by antocuni** [comment]
Thanks to all for the good feedback!
I think we all agree that we need to refactor/simplify the life cycle of the page. We agree less on the exact shape of JS plugins, which is hard to predict because we cannot anticipate what we will need until we have real use cases.

So, unless there are major objections I will proceed as follows:
1. start a PoC to refactor the life cycle of the page roughly following my proposal above
2. try to keep the number of hooks at minimum (we can add more later if needed)
3. do NOT advertise this as the official JS plugin API
4. if we are lucky, this will also allow us to kill svelte stores
5. see how things go, learn from mistakes, goto (1) and iterate

**Comment by pauleveritt** [comment]
> > My two cents: I agree with @fpliger ... I think the lifecycle on the JS should be the minimum.
> 
> I don't understand: what do you mean with that?

tl;dr I spent 12 hours last weekend writing a prototype and making a video related to this.

From the perspective of somebody writing something -- either a plugin, app, or just a page -- the "world" should be in Python. They shouldn't write part of their shareable thing in JS and part in Python.

Last weekend I wrote a prototype where `<py-counter>` was 100% in Python as a PyPI-sharable "component". It auto-generated a custom element class which could receive attributes, events, and mutations from the JS side. It just returned a string of HTML which morphdom used to update the DOM.

> > I confess to a secret agenda: I'm hoping to prove we can write custom elements completely in Python. I recognize that's a colossal change. I doubt PyScript will look at it and I'll do it as a Pyodide app. But in that mode, you want the "source of truth" in the interpreter.
> 
> personally I'm happy to move more logic to Python but we need to be aware of the tradeoffs:
> 
> 1. at this moment in time, python is slower (might be good enough for some cases, not for others)

Can you name something a plugin develop would want to do that doesn't rely on Python being initialized?

IMO, plugins that provide something useful *in the absence of Pyodide* are non-existent. Everything in PyScript is part of getting Python to execute something.

> 2. there are things which you cannot do in Python because you want to do them _before_ the runtime is initialized

Those things are internal housekeeping. Can you name some things that an external contributor would want to share with other people, that come under this heading?

> > Now, a more constructive response to @antocuni and his work. grinning I have "Diary Driven Development" showing your 1-5 ("load config" to "wait until the runtime is ready") in a dirt-simple fashion:
> > 
> > * Have a well-known data attribute added on startup to the body `<body data-pslifescyle>`
> > * As things change, change the value
> > * Have Mutation Observers that watch this and act
> 
> I like this idea, but it solves only a fraction of the issue. This is basically a pub/sub mechanism which is good e.g. for the splashscreen but it doesn't allow an easy way to sync between the various activities (e.g.: I want to wait until all the plugins have initialized the `runtime`)

All the plugins are loaded in the Python runtime. Each app registers in a Python registry. I have a video showing it. 😀

I think making PyScript plugin authors author on both the JS side and the Python side is a bad idea.

**Comment by antocuni** [comment]
> Last weekend I wrote a prototype where `<py-counter>` was 100% in Python as a PyPI-sharable "component". It auto-generated a custom element class which could receive attributes, events, and mutations from the JS side. It just returned a string of HTML which morphdom used to update the DOM.

I agree that python plugins should be first class citizens and people should be encouraged to use them when possible and when it makes sense.

I think that there is also a misunderstanding about the meaning of "plugins". You seem to be talking about "widgets", which in my view are just a subset of possible plugins (see below).

As I already said, in this proposal "plugins" are modeled after the successful model of pytest, in which plugins can use a set of standard hooks to intercept and modify in many ways the behavior of pytest itself: so they are much broader than just widgets.

> Can you name something a plugin develop would want to do that doesn't rely on Python being initialized?
> 
> IMO, plugins that provide something useful _in the absence of Pyodide_ are non-existent. Everything in PyScript is part of getting Python to execute something.

Random list of potential plugins which don't rely on pyodide (some real, some realistic, some just invented but to give an idea of what could be done):
- a micropython plugin and in general, plugins which offer alternative execution runtime. The pyodide runtime itself is exposed by a plugin which is distributed with pyscript and enabled by default, but it is perfectly reasonable to provide more and/or to disable pyodide
- the splashscreen: it *must* be displayed before pyodide is loaded, because that's exactly its job
- `<py-repl>`: it obviously relies on existing JS libraries for the editor part, it is probably easier to just write it in JS than trying to use them from Python. Moreover, we probably want to style and display the editor *before* pyodide is loaded, else the page looks *very* ugly for a few seconds
- in general, anything which needs to be styled before the page is loaded
- a `github` plugin which modifies the emscripten virtual filesystem to allow people to import and read files directly from github
- a `terminal` plugin which intercepts `stdout` and `stderr` to display them in a nice terminal on the page (it must be done at a lower level than pyodide)
- a `graphical debugger` plugin which allows to execute python code step-by-step. It obviously cannot be done in Python because the python engine is stopped during the execution

> > 2. there are things which you cannot do in Python because you want to do them _before_ the runtime is initialized
> 
> Those things are internal housekeeping. Can you name some things that an external contributor would want to share with other people, that come under this heading?

see above.

> I think making PyScript plugin authors author on both the JS side and the Python side is a bad idea.

I don't really understand this dogmatic position. In the browser, JS plays a role which is very similar to what C plays in a unix-like system. Everybody agrees that it's nice to write Python, but there are things which must be done in C, and it works very well.

Anyway, as I said above, we don't have to advertise the JS API for writing plugins right now. I think it's a better idea to play with it internally and then decide what to do.

**Comment by pauleveritt** [comment]
> > I think making PyScript plugin authors author on both the JS side and the Python side is a bad idea.
> 
> I don't really understand this dogmatic position. In the browser, JS plays a role which is very similar to what C plays in a unix-like system. Everybody agrees that it's nice to write Python, but there are things which must be done in C, and it works very well.

Point taken. There are some bootstrapping and machinery things that are reasonable to make extensible. And indeed, I was more focused on the 95% case of widgets, casual folks making things they want to share with other people.

I'll take my prototype and video, convert it to a "Pyodide Custom Elements" blog series or something, just as a teaching exercise.

**Comment by antocuni** [comment]
> Point taken. There are some bootstrapping and machinery things that are reasonable to make extensible. And indeed, I was more focused on the 95% case of widgets, casual folks making things they want to share with other people.

I think that you are correct here: I imagine that the vast majority of plugins will be custom widgets/components and we will have to make sure to fully support this use case. Thanks for pointing it out!

**Comment by marimeireles** [comment]
I don't really want to read the whole discussion 😅
@antocuni was this completed and the issue is now obsolete?

**Comment by fpliger** [comment]
@antocuni @marimeireles , I'm going to close this one because I think we covered [most of] it and changes should probably come up from a new and cleaner issue (also even just to keep things small and understandable).... Feel free to reopen if you disagree
